File: 0846-hand-of-straights/0846-hand-of-straights.java
class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        Map<Integer, Integer>mp = new TreeMap<>();
        for(int x:hand) {
            mp.put(x, mp.getOrDefault(x, 0) + 1);
        }
        for(int val:mp.keySet()) {
            int freq = mp.get(val);
            if(freq > 0) {
                for(int i=0; i<groupSize; i++) {
                    if(!mp.containsKey(val + i)) return false;
                    int nextVal = mp.get(val + i);
                    if(nextVal - freq < 0) return false;
                    mp.put(val + i, nextVal - freq);
                }
            }
        }
        return true;
    }
}

File: 0846-hand-of-straights/0846-hand-of-straights.cpp
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        map<int, int>mp;
        for(int x:hand) mp[x]++;
        for(auto it:mp) {
            int freq = it.second, val = it.first;
            if(freq > 0) {
                for(int i=0; i<groupSize; i++) {
                    mp[val + i] -= freq;
                    if(mp[val + i] < 0) return false;
                }
            }
        }
        return true;
    }
};